File: src/packages/agent_core/builtin_tools/noop.py
from __future__ import annotations

import time
from typing import Any

from packages.agent_core.executor import (
    ToolExecutionContext,
    ToolExecutionError,
    ToolExecutionResult,
)
from packages.agent_core.tools import ToolSpec as AgentToolSpec
from packages.llm_gateway import ToolSpec as LlmToolSpec

_ERROR_CLASS_ARGS_INVALID = "tool.args_invalid"

NOOP_AGENT_TOOL_SPEC = AgentToolSpec(
    name="noop",
    version="1",
    description="无副作用的空操作",
    risk_level="low",
)

NOOP_LLM_TOOL_SPEC = LlmToolSpec(
    name="noop",
    description="无副作用的空操作",
    json_schema={
        "type": "object",
        "properties": {},
        "additionalProperties": False,
    },
)


class NoopToolExecutor:
    async def execute(
        self,
        *,
        tool_name: str,
        args: dict[str, Any],
        context: ToolExecutionContext,
        tool_call_id: str | None = None,
    ) -> ToolExecutionResult:
        _ = (tool_name, context, tool_call_id)
        started = time.monotonic()
        if args:
            return ToolExecutionResult(
                error=ToolExecutionError(
                    error_class=_ERROR_CLASS_ARGS_INVALID,
                    message="noop 不接受参数",
                    details={"unexpected_fields": sorted(args.keys())},
                ),
                duration_ms=_duration_ms(started),
            )
        return ToolExecutionResult(result_json={"ok": True}, duration_ms=_duration_ms(started))


def _duration_ms(started: float) -> int:
    elapsed = time.monotonic() - started
    millis = int(elapsed * 1000)
    return millis if millis >= 0 else 0


__all__ = ["NOOP_AGENT_TOOL_SPEC", "NOOP_LLM_TOOL_SPEC", "NoopToolExecutor"]

File: src/services/worker_go/internal/tools/builtin/noop.go
package builtin

import (
	"context"
	"sort"
	"time"

	"arkloop/services/worker_go/internal/llm"
	"arkloop/services/worker_go/internal/tools"
)

const noopErrorArgsInvalid = "tool.args_invalid"

var NoopAgentSpec = tools.AgentToolSpec{
	Name:        "noop",
	Version:     "1",
	Description: "无副作用的空操作",
	RiskLevel:   tools.RiskLevelLow,
}

var NoopLlmSpec = llm.ToolSpec{
	Name:        "noop",
	Description: stringPtr("无副作用的空操作"),
	JSONSchema: map[string]any{
		"type":                 "object",
		"properties":           map[string]any{},
		"additionalProperties": false,
	},
}

type NoopExecutor struct{}

func (NoopExecutor) Execute(
	ctx context.Context,
	toolName string,
	args map[string]any,
	_ tools.ExecutionContext,
	_ string,
) tools.ExecutionResult {
	_ = ctx
	_ = toolName
	started := time.Now()

	if len(args) > 0 {
		fields := make([]string, 0, len(args))
		for key := range args {
			fields = append(fields, key)
		}
		sort.Strings(fields)
		return tools.ExecutionResult{
			Error: &tools.ExecutionError{
				ErrorClass: noopErrorArgsInvalid,
				Message:    "noop 不接受参数",
				Details:    map[string]any{"unexpected_fields": fields},
			},
			DurationMs: durationMs(started),
		}
	}

	return tools.ExecutionResult{
		ResultJSON: map[string]any{"ok": true},
		DurationMs: durationMs(started),
	}
}
